client/service/src/storage_monitor.rs

```diff
@@ -0,0 +1,143 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{config::Configuration, error::Error};
+use futures::StreamExt;
+use nix::{errno::Errno, sys::statvfs::statvfs};
+use notify::{Config, Event, RecommendedWatcher, RecursiveMode, Watcher};
+use sc_utils::mpsc::{tracing_unbounded, TracingUnboundedReceiver};
+use std::path::{Path, PathBuf};
+
+const LOG_TARGET: &str = "storage-monitor";
+
+pub struct StorageMonitorService {
+	/// watched path
+	path: PathBuf,
+	/// notify's events receiver
+	stream: TracingUnboundedReceiver<Result<Event, notify::Error>>,
+	/// number of bytes that shall be free and available on the filesystem for watched path
+	threshold: u64,
+	/// keeps the ref for file system watcher
+	_watcher: RecommendedWatcher,
+}
+
+impl StorageMonitorService {
+	/// creates new StorageMonitorService for given client config
+	pub fn new_for_config(config: &Configuration) -> Result<Option<Self>, Error> {
+		Ok(match (config.available_storage_theshold, config.database.path()) {
+			(0, _) => {
+				log::info!(
+					target: LOG_TARGET,
+					"StorageMonitorService: threshold 0 given, storage monitoring disabled",
+				);
+				None
+			},
+			(_, None) => {
+				log::warn!(
+					target: LOG_TARGET,
+					"StorageMonitorService: no database path to observe",
+				);
+				None
+			},
+			(threshold, Some(path)) => {
+				let (sink, stream) = tracing_unbounded("mpsc_storage_monitor", 1024);
+
+				let mut watcher = RecommendedWatcher::new(
+					move |res| {
+						sink.unbounded_send(res).unwrap();
+					},
+					Config::default(),
+				)
+				.map_err(|e| Error::Other(format!("Could not create fs watcher {e}")))?;
+
+				watcher
+					.watch(path.as_ref(), RecursiveMode::Recursive)
+					.map_err(|e| Error::Other(format!("Could not start fs watcher {e}")))?;
+
+				log::debug!(
+					target: LOG_TARGET,
+					"Initializing StorageMonitorService for db path: {:?}",
+					path,
+				);
+
+				Self::check_free_space(&path, threshold)?;
+
+				Some(StorageMonitorService {
+					path: path.to_path_buf(),
+					stream,
+					threshold,
+					_watcher: watcher,
+				})
+			},
+		})
+	}
+
+	/// main monitoring loop, intended to be spawn as essential task
+	pub async fn run(mut self) {
+		while let Some(watch_event) = self.stream.next().await {
+			match watch_event {
+				Ok(_) =>
+					if Self::check_free_space(&self.path, self.threshold).is_err() {
```

Comment:
Events from fs-watcher are sometimes reported more often then 1 per second. Some throttling can be added here to avoid excessive calls to `statvfs`. 

Log sample:
```
2023-01-05 18:56:39.007 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:39.007  INFO tokio-runtime-worker substrate: ✨ Imported #4 (0xff57…164a)    
2023-01-05 18:56:39.008 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:39.008 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:39.416 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:39.419 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:39.582 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:40.085 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:40.751 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:40.754 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:40.754 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:40.842 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:40.842 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:41.420 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:42.003  INFO tokio-runtime-worker sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0xff578913a9cfb414015a04cd13cd14d2c5b07ab46c869f5b80dc7aaff7c1164a    
2023-01-05 18:56:42.004  INFO tokio-runtime-worker sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 5 (0 ms) [hash: 0x4bb985888284217d7dd00d478396f57ebb667977a17b698af1ed3889cfda0717; parent_hash: 0xff57…164a; extrinsics (1): [0x767b…e13a]]    
2023-01-05 18:56:42.005  INFO tokio-runtime-worker babe: 🔖 Pre-sealed block for proposal at 5. Hash now 0xfb06f0bb33dfef41d2492e3f8a2dc74b51d4ccdb483806566676979ff01d455f, previously 0x4bb985888284217d7dd00d478396f57ebb667977a17b698af1ed3889cfda0717.    
2023-01-05 18:56:42.006  INFO tokio-runtime-worker substrate: ✨ Imported #5 (0xfb06…455f)    
2023-01-05 18:56:42.006 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:42.006 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:42.006 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:42.086 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:42.088 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:42.345 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:42.345 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:42.407  INFO tokio-runtime-worker substrate: 💤 Idle (0 peers), best: #5 (0xfb06…455f), finalized #2 (0x5d6f…3153), ⬇ 0 ⬆ 0    
2023-01-05 18:56:42.754 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:43.420 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:43.423 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:43.681 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:43.681 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:44.089 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:44.755 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:44.757 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:45.001  INFO tokio-runtime-worker sc_basic_authorship::basic_authorship: 🙌 Starting consensus session on top of parent 0xfb06f0bb33dfef41d2492e3f8a2dc74b51d4ccdb483806566676979ff01d455f    
2023-01-05 18:56:45.002  INFO tokio-runtime-worker sc_basic_authorship::basic_authorship: 🎁 Prepared block for proposing at 6 (0 ms) [hash: 0xb2e83cf3b9c2387fde0c332c1fb9ee5f8d4621ac0e421669f6efd7861f8155a3; parent_hash: 0xfb06…455f; extrinsics (1): [0x9616…8461]]    
2023-01-05 18:56:45.004  INFO tokio-runtime-worker babe: 🔖 Pre-sealed block for proposal at 6. Hash now 0x89244954781773e6d0b85656f776f08e0a8a435131695e5acd1504a03dd19422, previously 0xb2e83cf3b9c2387fde0c332c1fb9ee5f8d4621ac0e421669f6efd7861f8155a3.    
2023-01-05 18:56:45.005 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:45.005  INFO tokio-runtime-worker substrate: ✨ Imported #6 (0x8924…9422)    
2023-01-05 18:56:45.005 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:45.005 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:45.333 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:45.333 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:45.423 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:46.090 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:46.092 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:46.092 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:46.309 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:46.309 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000    
2023-01-05 18:56:46.759 TRACE tokio-runtime-worker storage-monitor: free:808564 , threshold:270000  
```

Reply:
From `notify::Watcher` docs it looks like every inotify event is reported, like every read access to files under the path. Probably we don't need to call `statvfs` for every such event and can throttle to something like 1 Hz. May be we don't even need inotify and can just measure free space every 1-2 secs.

Reply:
periodic polling added, notify removed: https://github.com/paritytech/substrate/pull/13082/commits/86197668dcc5efd8c003fa1042f1e25d810da1ef,